docs/sourcemaps.asciidoc

```diff
@@ -1,38 +1,192 @@
+* `sourcemap` - The location of the source map.
+If you have multiple source maps, you'll need to upload each individually.
 
-[[sourcemap-apply]]
 [float]
-== How source maps are applied
+[[sourcemap-curl]]
+==== Upload via curl
+
+Here’s an example curl request that uploads the source map file created in the previous step:
 
-APM Server needs to find the correct source map for every `stack trace frame` in an event.
-To do so, it tries the following:
+[source,console]
+----
+SERVICEVERSION=`node -e "console.log(require('./package.json').version);"` && \ <1>
+curl http://localhost:8200/assets/v1/sourcemaps -X POST \
+    -F sourcemap="@./dist/app.min.js.map" \
+    -F service_version="$SERVICEVERSION" \
+    -F bundle_filepath="http://localhost/app.min.js" \
+    -F service_name="myService"
+----
+<1> This example uses the version from `package.json`.
 
-* Compare the event's `service.name` with the source map's `service_name`
-* Compare the event's `service.version` with the source map's `service_version`
-* Compare the stack trace frame's `abs_path` with the source map's `bundle_filepath`
+Add a <<secret-token,secret token>> to your curl command using an Authorization header:
 
-If a source map is found, the following attributes in the `stack trace frames` are overwritten:
+[source,console]
+----
+-H "Authorization: Bearer <token>"
+----
+
+Or, add an <<api-key,API key>> to your curl command using an Authorization header:
+
+[source,console]
+----
+-H "Authorization: ApiKey <token>"
+----
+
+[float]
+[[sourcemap-custom-app]]
+==== Upload via a custom app
+
+Here’s an example Node.js application that uploads the source map file created in the previous step:
+
+[source,js]
+----
+console.log('Uploading sourcemaps!')
+var request = require('request')
+var filepath = './dist/app.min.js.map'
+var formData = {
+  sourcemap: fs.createReadStream(filepath),
+  service_version: require("./package.json").version, // Or use 'git-rev-sync' for git commit hash
+  bundle_filepath: 'http://localhost/app.min.js',
+  service_name: 'service-name’ <1>
+}
+request.post({url: 'http://localhost:8200/assets/v1/sourcemaps',formData: formData}, function (err, resp, body) {
+  if (err) {
+    console.log('Error while uploading sourcemaps!', err)
+  } else {
+    console.log('Sourcemaps uploaded!')
+  }
+})
+----
+<1> `service_name` should match the service name specified when initializing the RUM Agent
+
+[float]
+[[sourcemap-done]]
+=== Next steps
 
-* `filename`
-* `function`
-* `line number`
-* `column number`
-* `abs path` is https://golang.org/pkg/path/#Clean[cleaned] to be the shortest path name equivalent to the given path name.
+That's it! The source map has been uploaded to Elasticsearch,
+and exception stack traces should now be correctly mapped to your source code.
 
-If multiple source maps are found,
-the one with the latest upload timestamp is used.
+More information:
```

Comment:
Somewhere it could be a good idea to nudge users to upload sourcemaps as part of their deployment procedure, and not when start seeing errors with random gibberish, because uploading sourcemaps after those errors happen won't make them readable (errors must occur again)

Reply:
Added to the top of the page